File: commandAndMessage/command/CommandHandler.java
package commandAndMessage.command;

import com.sun.javafx.UnmodifiableArrayList;

import java.util.List;
import java.util.Optional;

/**
 * Created by Mark Chimes on 2017/11/05.
 */
public abstract class CommandHandler implements CommandOrMessageIterator {
    protected CommandIterator iterator;

    public String currentText() {
        return iterator.currentText();
    }

    public int currentIndex() {
        return iterator.currentIndex();
    }

    protected void setTextOrItemIterator(CommandIterator iterator) {
        this.iterator = iterator;
    }

    public Optional<CommandHandler> nextCommand() {
        CommandHandler handler = currentCommand().nextCommand;

        if (handler instanceof NullCommandHandler) {
            return Optional.empty();
        } else {
            return Optional.of(handler);
        }
    }

    public List<String> getNextMessage() {
        return currentCommand().nextMessage;
    }

    public List<String> getHelpText() {
        return currentCommand().helpText;
    }

    protected void setIteratorCommands(List<CommandTuple> commands) {
        setTextOrItemIterator(new CommandIterator(commands));
    }

    public final CommandTuple currentCommand() {
        return iterator.currentCommand();
    }

    public void performPreviousInListCommand() {
        iterator.performPreviousInListCommand();
    }

    public void performNextInListCommand() {
        iterator.performNextInListCommand();
    }

    public boolean isClearingCommandStack() {
        return currentCommand().shouldClearCommandStack;
    }

    public void performConfirmCommand() {
        currentCommand().executable.run();
    };
}

File: commandAndMessage/command/CommandHandler.kt
package commandAndMessage.command

import com.sun.javafx.UnmodifiableArrayList
import java.util.Optional

/**
 * Created by Mark Chimes on 2017/11/05.
 */
abstract class CommandHandler : CommandOrMessageIterator {
    lateinit var iterator: CommandIterator

    override fun currentText(): String {
        return iterator.currentText()
    }

    override fun currentIndex(): Int {
        return iterator.currentIndex()
    }

    protected fun setTextOrItemIterator(iterator: CommandIterator) {
        this.iterator = iterator
    }

    fun nextCommand(): Optional<CommandHandler> {
        val handler = currentCommand().nextCommand

        return if (handler is NullCommandHandler) {
            Optional.empty()
        } else {
            Optional.of(handler)
        }
    }

    val nextMessage: List<String>
        get() = currentCommand().nextMessage

    open val helpText: List<String>
        get() = currentCommand().helpText

    protected fun setIteratorCommands(commands: List<CommandTuple>) {
        setTextOrItemIterator(CommandIterator(commands))
    }

    fun currentCommand(): CommandTuple {
        return iterator.currentCommand()
    }

    override fun performPreviousInListCommand() {
        iterator.performPreviousInListCommand()
    }

    override fun performNextInListCommand() {
        iterator.performNextInListCommand()
    }

    val isClearingCommandStack: Boolean
        get() = currentCommand().shouldClearCommandStack

    fun performConfirmCommand() {
        currentCommand().executable?.run()
    }
}